nixos-modules/microvm/options.nix

```diff
@@ -559,6 +575,65 @@ in
       '';
     };
 
+    qemu.pcieRootPorts = mkOption {
+      description = ''
+        A list of PCIe root ports that can be used for hot-plugging PCIe devices.
+        This is particularly useful on the Q35 machine type, which does not support
+        hot-plugging on the base PCIe root bus (pcie.0). Creating root ports allows
+        attaching and detaching PCIe devices at runtime and can also be useful for
+        devices that require their own dedicated PCIe slot with a fixed address, etc.
+        For additional details see the QEMU PCI Express Guidelines:
+        <https://gitlab.com/qemu-project/qemu/-/blob/master/docs/pcie.txt>
+      '';
+      default = [];
+      example = literalExpression /* nix */ ''
+        [ {
+          bus = "pcie.0";
+          id = "pci_port_0";
+          chassis = 0;
+        }]
+      '';
+      type = with types; listOf (submodule {
```

Comment:
If the entire type can be a list of null, then each individual option does not need to be nullable, unless each option can be set individually.

Reply:
Thanks, good to know. I think it doesn’t really make sense to allow all parameters to be null because the ports wouldn’t be usable without id so I made it non-null.